x/ccv/provider/keeper/validator_set_update.go

```diff
@@ -0,0 +1,187 @@
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.ConsumerValidatorKey(chainID, providerAddr.ToSdkConsAddr())) != nil
+}
+
+// GetConsumerValSet returns all the consumer validators for chain `chainID`
+func (k Keeper) GetConsumerValSet(
+	ctx sdk.Context,
+	chainID string) (validators []types.ConsumerValidator) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.ChainIdWithLenKey(types.ConsumerValidatorBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, key)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		iterator.Value()
+		var validator types.ConsumerValidator
+		if err := validator.Unmarshal(iterator.Value()); err != nil {
+			panic(fmt.Errorf("failed to unmarshal ConsumerValidator: %w", err))
+		}
+		validators = append(validators, validator)
+	}
+
+	return validators
+}
+
+// ComputeNextEpochConsumerValSet returns the next validator set that is responsible for validating consumer
+// chain `chainID`, based on the bonded validators.
+func (k Keeper) ComputeNextEpochConsumerValSet(
+	ctx sdk.Context,
+	chainID string,
+	bondedValidators []stakingtypes.Validator,
+) []types.ConsumerValidator {
+	var nextValidators []types.ConsumerValidator
+	for _, val := range bondedValidators {
+		// get next voting power and the next consumer public key
+		nextPower := k.stakingKeeper.GetLastValidatorPower(ctx, val.GetOperator())
+		consAddr, err := val.GetConsAddr()
+		if err != nil {
+			// this should never happen but is recoverable if we exclude this validator from the `nextValidators`
+			k.Logger(ctx).Error("could not get consensus address of validator",
+				"validator", val.GetOperator().String(),
+				"error", err)
+			continue
+		}
+		nextConsumerPublicKey, foundConsumerPublicKey := k.GetValidatorConsumerPubKey(ctx, chainID, types.NewProviderConsAddress(consAddr))
+		if !foundConsumerPublicKey {
+			// if no consumer key assigned then use the validator's key itself
+			k.Logger(ctx).Info("could not retrieve public key for validator on consumer chain because"+
+				" the validator did not assign a new consumer key",
+				"validator", val.GetOperator().String(),
+				"chainID", chainID)
+			nextConsumerPublicKey, err = val.TmConsPublicKey()
+			if err != nil {
+				// this should never happen and might not be recoverable because without the public key
+				// we cannot generate a validator update
+				panic(fmt.Errorf("could not retrieve validator's (%+v) public key: %w", val, err))
+			}
+		}
+
+		nextValidator := types.ConsumerValidator{
+			ProviderConsAddr:  consAddr,
+			Power:             nextPower,
+			ConsumerPublicKey: &nextConsumerPublicKey,
+		}
+		nextValidators = append(nextValidators, nextValidator)
+	}
+
+	return nextValidators
+}
+
+// DiffValidators compares the current and the next epoch's consumer validators and returns the `ValidatorUpdate` diff
+// needed by CometBFT to update the validator set on a chain.
+func DiffValidators(
```

Comment:
Nit: I think you have slight extra complexity here because you index your "sets" by providerConsAddr, which you actually never use. afaict you could simplify this if you make your isCurrentValidator and isNextValidator maps from ConsumerPublicKey to ConsumerValidator (note you never use the valProviderConsAddr except for as the map key), then I think it becomes conceptually a bit easier and you get rid of the special case where a validator changed consumer keys

Reply:
Great idea! This indeed simplifies `DiffValidators`. I updated the PR.